File: python/leetcode/combination-sum.py
class Solution:
    def combinationSum(self, c, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        if target < 0:
            return []
        if target == 0:
            return [[]]
        result = []
        for i in range(len(c)):
            part_sum = self.combinationSum(c[i:], target - c[i])
            for s in part_sum:
                result.append(s + [c[i]])
        return result

File: java/leetcode/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] c, int target) {
        List<List<Integer>> result = new ArrayList();
        if (target < 0) {
            return result;
        }
        if (target == 0) {
            List<Integer> emp = new ArrayList();
            result.add(emp);
            return result;
        }
        for (int i = 0; i < c.length; i++) {
            List<List<Integer>> partSum = combinationSum(Arrays.copyOfRange(c, i, c.length), target - c[i]);
            for (List<Integer> s : partSum) {
                List<Integer> newList = new ArrayList();
                newList.add(c[i]);
                newList.addAll(s);
                result.add(newList);
            };
        }
        return result;
    }
}
